packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -0,0 +1,648 @@
+       textEditingController = textEditingController ?? TextEditingController() {
+    this.textEditingController.addListener(_onChangedField);
+  }
+
+  final bool _ownsTextEditingController;
+
+  /// All possible options that can be selected.
+  ///
+  /// If left null, a custom [getResults] method must be provided that handles
+  /// generating some results based on the text in the field.
+  final List<T> options;
+
+  /// The [TextEditingController] that represents the field.
+  final TextEditingController textEditingController;
+
+  /// A function that returns the possible results given the text in the field.
+  ///
+  /// If [options] is null, then this field must not be null. This may be the
+  /// case when querying an external service for results, for example.
+  ///
+  /// Defaults to a simple string-matching filter of [options].
+  final AutocompleteResultsGetter<T> getResults;
+
+  /// Returns the string to display in the field when the option is selected.
+  ///
+  /// This is useful when using a custom T type for AutocompleteController and
+  /// the string to display is different than the string to search by.
+  ///
+  /// If not provided, will use `option.toString()`.
+  ///
+  /// See also:
+  ///   * [filterStringForOption], which can be used to specify a custom string
+  ///     to filter by.
+  final _AutocompleteOptionToString<T> displayStringForOption;
+
+  /// Returns the string to match against when filtering the given option.
+  ///
+  /// This is useful when using a custom T type for AutocompleteController and
+  /// the string to display is different than the string to search by. This is
+  /// only used when [getResults] is null and the default getResults filter is
+  /// used.
+  ///
+  /// If not provided, will use `option.toString()`.
+  ///
+  /// See also:
+  ///   * [displayStringForOption], which can be used to specify a custom String
+  ///     to be shown in the field.
+  final _AutocompleteOptionToString<T> filterStringForOption;
+
+  /// The current results being returned by [getResults].
+  ///
+  /// This is a [ValueNotifier], so it can be listened to for changes.
+  final ValueNotifier<List<T>> results = ValueNotifier<List<T>>(<T>[]);
+
+  // The default way to convert an option to a string.
+  static String _defaultStringForOption<T>(T option) {
+    return option.toString();
+  }
+
+  /// Clean up memory created by the AutocompleteController.
+  ///
+  /// Call this when the AutocompleteController is no longer needed, such as in
+  // the dispose method of the widget it was created in.
+  void dispose() {
+    textEditingController.removeListener(_onChangedField);
+    if (_ownsTextEditingController) {
+      textEditingController.dispose();
+    }
+  }
+
+  // Called when textEditingController reports a change in its value.
+  void _onChangedField() {
+    final List<T> resultsValue = getResults == null
+        ? _filterByString(textEditingController.value.text)
```

Comment:
Agreed about passing the whole TextEditingValue.

As it is now, it will filter by whatever is in `text`, so that includes composing text.  I just went down a big rabbit hole on this, but I think this existing behavior is correct for now, explained below.

There are two main use cases that I can think of:

  1. Using a composing keyboard to filter a list of non-Latin strings (e.g. using a pinyin keyboard and filtering a list of Chinese provinces written in Chinese characters).
  2. Using a composing keyboard to filter a list of Latin strings (e.g. using a pinyin keyboard and filtering a list of U.S. states in English).

In case 1, it's suboptimal that typing a composing character will filter everything out of the list. However, really the app developer should be handling this case on their own.  Using the tools we give them like filterStringForOption or getResults, the app developer could filter based on both the Chinese characters and the pinyin, even though only the Chinese characters are displayed in the results list.  This isn't something that Flutter itself can really do without bringing a big Chinese dictionary into the project (plus every other language we support that using composing keyboards).

In case 2, correct me if I'm wrong, but I would think that users would want the composing characters to filter the list of states as they type.

Let me know if I'm off on this!